sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+
+        let mut console_tty: Option<PathBuf> = None;
+        let mut spi_interface: Option<BulkInterface> = None;
+        let mut uart_ttys: HashMap<String, PathBuf> = HashMap::new();
+
+        let config_desc = device.active_config_descriptor()?;
+        // Iterate through each USB interface, discovering e.g. supported UARTs.
+        for interface in config_desc.interfaces() {
+            for interface_desc in interface.descriptors() {
+                if let Some(idx) = interface_desc.description_string_index() {
+                    if let Ok(interface_name) = device.read_string_descriptor_ascii(idx) {
+                        let ports = device.port_numbers()?.iter().map(
+                            |id| id.to_string()).collect::<Vec<String>>().join(".");
+                        let interface_path = path
+                            .join(format!("{}-{}", device.bus_number(), ports))
+                            .join(format!("{}-{}:{}.{}",device.bus_number(), ports,
+                                          config_desc.number(), interface.number()));
+                        // Check the ASCII name of this USB interface.
+                        if interface_name == "HyperDebug Shell" {
+                            // We found the "main" control interface of HyperDebug, allowing
+                            // textual commands to be sent, to e.g. manipoulate GPIOs.
+                            console_tty = Some(Self::find_tty(&interface_path)?)
+                        } else if interface_name.starts_with("UART") {
+                            // We found an UART forwarding USB interface.
+                            uart_ttys.insert(
+                                interface_name.clone(),
+                                Self::find_tty(&interface_path)?);
+                        } else if interface_name == "SPI" {
+                            // We found the SPI forwarding USB interface (this one interface
+                            // allows multiplexing physical SPI ports.)
+                            let mut in_endpoint: u8 = 0;
+                            let mut out_endpoint: u8 = 0;
+                            for endpoint_desc in interface_desc.endpoint_descriptors() {
+                                if endpoint_desc.transfer_type() == rusb::TransferType::Bulk {
+                                    match endpoint_desc.direction() {
+                                        rusb::Direction::In => {
+                                            in_endpoint = endpoint_desc.address();
+                                        }
+                                        rusb::Direction::Out => {
+                                            out_endpoint = endpoint_desc.address();
+                                        }
+                                    }
+                                }
+                            }
+                            if in_endpoint == 0 || out_endpoint == 0 {
```

Comment:
Maybe `Option`s are more appropriate for this?

Reply:
Done